### ISSUE TITLE
Incompatibility: MUI Sleep

### ISSUE THREAD

**Issue by deadw00d** [deadw00d/AROS#51]
"One other problem/feature I noticed on AROS.  AROS has different semantics w.r.t MUIA_Application_Sleep and newly created windows.

There are 2 differences which I in the following I will call open-window-only and sleep-underflow

open-window-only: In classic MUI (m68k, AOS4, MorphOS) one can create a new window, set the application to sleep and then open the new window, which will not be put to sleep (but all other windows are).  On AROS though the newly created window is also put to sleep.  Classic MUI only sets open windows to sleep whereas AROS seems to also set hidden/closed windows to sleep.

It is easy to work-around this by first setting the application to sleep before creating and opening the new window.

sleep-underflow Classic MUI allows to unsleep the application and windows that weren't put to sleep are still in their no-sleep state.  Said differently, the sleep counter of a window does not drop below 0 when calling MUIA_Application_Sleep, FALSE.  On AROS though it seems this sleep counter underflows, thus when creating and opening a new window and unsleeping the application, the new window is set to sleep.  It gets even more weird when calling MUIA_Application_Sleep, TRUE as this can make a window to get a zero sleep counter and allow user input.

It is possible to work around this by making sure that between every call to MUIA_Application_Sleep, TRUE and MUIA_Application_Sleep, FALSE, no new window is created.  But it is not a trivial one.  Often it is just very convenient to set the whole application to sleep, do whatever changes are necessary, which might include adding new windows, and then unsleep the application.

I just did an audit of all sleep/unsleep calls in MUIbase in order to match the semantics of AROS (see e.e.g this commit: https://sourceforge.net/p/muibase/code/3918).

MUIbase is probably not the only application that runs into this issue, thus, my recommendation would be to change the semantics to how classic MUI does it.  And BTW, both GTK and Qt also work the same way as classic MUI.  In particular the sleep-underflow change would be highly desirable."

https://ae.amigalife.org/index.php?topic=823.msg9337#msg9337

**Comment by steffengutmann** [comment]
Attached are two test cases with commands on how to reproduce and outputs on m68k/MUI3.8 and i386-aros.

I think MUI3.8 has its own sleep counter for MUIA_Application_Sleep while AROS seems to re-use the MUIA_Window_Sleep one.  This is probably part of the problem.

[sleep-open-window-only.c.txt](https://github.com/deadw00d/AROS/files/7790416/sleep-open-window-only.c.txt)
[sleep-underflow.c.txt](https://github.com/deadw00d/AROS/files/7790418/sleep-underflow.c.txt)

**Comment by steffengutmann** [comment]
Thinking more about this, I think the real issue here is that AROS re-used the MUIA_Window_Sleep counter for MUIA_Application_Sleep while classic MUI seems to use a separate counter for MUIA_Application_Sleep.

And I think it makes sense to have separate sleep counters.  Here is how I would implement the application and window sleep counters:

On MUIA_Application_Sleep, the application-sleep counter is updated. If it is non-zero then all open windows are set to busy.  If it is zero then all open windows with a zero window-sleep counter are set to non-busy.

On MUIA_Window_Sleep, the window-sleep counter is updated.  If it is non-zero then the window it set to busy.  If it is zero and the application-sleep counter is also zero, then the window is set to non-busy.

This would match the classic MUI behavior and fix both issues in the original comment.

**Comment by deadw00d** [comment]
Thanks for the suggestion - I'll use it once I get to solve this topic. If you do any additional tests or have more thoughts in the mean time, please share.

**Comment by steffengutmann** [comment]
It is also worth taking a look at the auto docs for [MUIA_Application_Sleep](https://github.com/amiga-mui/muidev/wiki/MUI_Application/#MUIA_Application_Sleep) where it says under Notes: _"Only opened windows will be put to sleep. However, leaving the sleep state will affect all windows regardless of the current open state."_ and [MUIA_Window_Sleep](https://github.com/amiga-mui/muidev/wiki/MUI_Window/#MUIA_Window_Sleep) where is written: _"Only opened windows can be put to sleep. However, currently closed windows can be taken out of sleep nevertheless."_

This means that my suggested logic should be refined such the all windows are considered when counters reach zero, not just open ones.

**Comment by steffengutmann** [comment]
Actually the proposed solution in [comment 3](https://github.com/deadw00d/AROS/issues/51#issuecomment-1002854888) would not work (even with the modification in the previous comment).  Here is a counter example:

```
OpenWindow(A)
ApplicationSleep(TRUE)  // count=1, Window A is busy
OpenWindow(B);  // Window B is non-busy
ApplicationSleep(TRUE)  // count=2, Windows A and B are busy
ApplicationSleep(FALSE)  // count=1, Windows A and B are still busy, but only Window A should be
ApplicationSleep(FALSE)  // count=0, Windows A and B are back to non-busy

```

This means, the applicatioin-busy-counter must be per window.  Thus I suggest the following instead: two sleep counters for each window: an app-busy-counter and a win-busy-counter.

On ApplicationSleep(TRUE) all open windows increase their app-sleep counter.
On ApplicationSleep(FALSE) all windows (whether open or closed) decrease their app-sleep counter.
On WindowSleep(TRUE) the win-sleep counter is increased if the window is open.
On WindowSleep(FALSE) the win-sleep counter is decreased (irrespective of whether the window is open or closed).

Both counters never drop below zero.

If any of app-sleep-counter or win-sleep-counter is non-zero, the window is set to busy, otherwise to non-busy.

**Comment by deadw00d** [comment]
So if I'm getting things right:
WindowSleep(TRUE) increases window counter by one.
WindowSleep(FALSE) decreases window counter by one.
ApplicationSleep(TRUE) is like broadcasting WindowSleep(TRUE) to all open windows.
ApplicationSleep(FALSE) is like broadcasting WindowSleep(FALSE) to all windows (opened and not).
?

**Comment by steffengutmann** [comment]
That sounds right, except that on Amiga/MUI ApplicationSleep is doing the broadcasting w/o calling WindowSleep, i.e. MUI seems to use a separate counter for each window in the ApplicationSleep method.

I attach a small modification to the original test case.  In the last test it becomes clear that Amiga/MUI is not re-using the WindowSleep counter.

[sleep-open-window-only.c.txt](https://github.com/deadw00d/AROS/files/7876285/sleep-open-window-only.c.txt)

**Comment by deadw00d** [comment]
If I read this last test correctly, setting Application_Sleep make the open window "physically sleep" (that is it is "busy/disable") but the Window_Sleep flag is set to FALSE?

Could you also point me to the cross-compiler package you are using for amiga-m68k?

**Comment by steffengutmann** [comment]
I looked into this a bit more, and the unit test I wrote were all wrong.  Sorry!

The wrong part in the unit tests posted previously here is that the created window does not have a root object.  Somehow MUI does not allow to open such a window.  Since the unit tests are very short and the GUI is never shown, I didn't realize that.

Here are new unit tests.  For sleep-open-window-only it shows that MUI is actually doing what you thought: MUIA_Application_Sleep broadcasts MUIA_Window_Sleep to all windows.  There is not a second sleep counter that I originally was suspecting.  This should make the fix really easy: in MUIA_Window_Sleep ([here](https://github.com/aros-development-team/AROS/blob/3b28655d041b42bcbf2526b1862fe2ca18b58cfe/workbench/libs/muimaster/classes/window.c#L3241)) only increase the counter when the window is open.

The sleep-underflow case is still the same, but I have also updated the unit test with a correctly created window.

[sleep-open-window-only.c.txt](https://github.com/deadw00d/AROS/files/7885297/sleep-open-window-only.c.txt)
[sleep-underflow.c.txt](https://github.com/deadw00d/AROS/files/7885298/sleep-underflow.c.txt)

I am using the gcc-3.4 cross compiler found [here](http://kas1e.mikendezign.com/zerohero_crosscompilers_backup/cross-compiler_os3.html).  I wish there were some newer  cross compiler though since that version has a lot of bugs when it comes to 64 bit integers.

**Comment by deadw00d** [comment]
Great! Thanks a lot for corrections and the cross-compiler. I'll adjust the unit tests and indeed this makes the fix easier.

**Comment by deadw00d** [comment]
I was able to install the cross-compiler and it seems it work (hello world works). I have a problem with MUI headers though. I installed the ones from Aminet (https://aminet.net/dev/mui/mui38dev.lha). When trying to compile sleep-underflow I'm getting these errors. Any idea what is wrong? What headers are you using?
`In file included from /usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/proto/muimaster.h:6,
                 from sleep-underflow.c:17:
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h: In function `MUI_AllocAslRequest':
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:32: error: asm-specifier for variable `res' conflicts with asm clobber list
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:32: error: asm-specifier for variable `d0' conflicts with asm clobber list
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:32: error: asm-specifier for variable `a0' conflicts with asm clobber list
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h: In function `MUI_AslRequest':
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:55: error: missing terminating " character
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: `jsr' undeclared (first use in this function)
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: (Each undeclared identifier is reported only once
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: for each function it appears in.)
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: syntax error before "a6"
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: stray '@' in program
/usr/local/amiga/lib/gcc/m68k-amigaos/3.4.0/../../../../m68k-amigaos/sys-include/inline/muimaster.h:56: error: missing terminating " character`

**Comment by steffengutmann** [comment]
Not sure what is wrong here, but likely the headers don't match.  I don't remember anymore where I got the headers from.  Have you tried the NDK 3.9 on the same link?  Otherwise, maybe disable inline code by using -DNO_INLINE (or check what alternatives proto/muimaster.h has to offer).

**Comment by deadw00d** [comment]
Fixed in d46d35730510d0d0602a46920341ea517cf90f7e
